Add Azure Authenticator Host Policy generator

abrahamko opened the pull request:
I followed @izgeri [comment on this issue](https://github.com/cyberark/conjur/issues/1497#issuecomment-617929172) to add a tool for generating Host policy that will comply with Azure authenticator.
She recommended me to add it to your tool here.

Using this tool one can put host name and azure token and get the necessary host definition in the policy that comply with azure authenticator.

abrahamko committed:
Add Azure Authenticator Host Policy generator

orenbm commented on views/welcome.html:
```diff
@@ -10,6 +10,7 @@ <h3>To get started, select a task below:</h3>
       <a class="list-group-item list-group-item-action" id="list-standby-list" data-toggle="list" href="#list-standby" role="tab" aria-controls="standby">Deploy Standby</a>
       <a class="list-group-item list-group-item-action" id="list-follower-list" data-toggle="list" href="#list-follower" role="tab" aria-controls="followers">Deploy Followers</a>
       <a class="list-group-item list-group-item-action" id="list-integration-list" data-toggle="list" href="#list-integration" role="tab" aria-controls="integration">Secure Apps</a>
+      <a class="list-group-item list-group-item-action" id="list-create-azure-host-policy-list" data-toggle="list" href="#list-create-azure-host-policy" role="tab" aria-controls="integration">Azure Authenticator</a>
```
The rest of the titles our of actions. I think we should change this one to "Generate Azure Authenticator host policy"

orenbm commented on views/welcome.html:
```diff
@@ -83,8 +84,22 @@ <h2>Integrations</h2>
+        <div class="row">
+          <div class="col-6">
+            <Br/>
+            <h2>Create Azure Host Policy</h2>
```
I would go with "Generate Azure Authenticator host policy"

orenbm commented on views/welcome.html:
```diff
@@ -83,8 +84,22 @@ <h2>Integrations</h2>
+          <div class="col-6">
+            <Br/>
+            <h2>Create Azure Host Policy</h2>
+            Generate the Host policy needed to authenticate using Azure Authenticator from Azure token<br/>
```
I would go with "Generate a policy defining a host that can authenticate with the Azure Authenticator, using an Azure AD token"

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+<h1>Azure Authenticator Policy Generator</h1>
```
Generate Azure Authenticator host policy

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+<h1>Azure Authenticator Policy Generator</h1>
+
+Generate Host Policy needed to use Azure Authenticator using Azure token
```
Generate a policy defining a host that can authenticate with the Azure Authenticator, using an Azure AD token

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+    <tbody>
+        <tr>
+            <th scope="row">1</th>
+            <td>Application ID</td>
```
host ID

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+            <td>Application ID</td>
+            <td><input type="text" id="txtAppId"></td>
+            <td>The ID you want to use in the policy to identify this application.
+            Used as host ID.</td>
```
this part is redundant

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+            <th scope="row">1</th>
+            <td>Application ID</td>
+            <td><input type="text" id="txtAppId"></td>
+            <td>The ID you want to use in the policy to identify this application.
```
... to identify the host.

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+
+<br />
+
+Azure access token
```
Azure AD token

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+        annotations:
+          authn-azure/subscription-id: {{subscription ID}}
+          authn-azure/resource-group: {{resource group}}
+          authn-azure/{{user or system}}-assigned-identity: {{identity}}
```
i would add `# Optional` at the and of the line

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+            output = "Application ID is missing";
+        }
+        else if (!encodedToken){
+            output = "Azure token is missing";
```
Azure AD token

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+        let appId = $("#txtAppId").val();
+        let encodedToken = $("#encodedToken").val();
+        if (!appId){
+            output = "Application ID is missing";
```
Host ID is missing

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+        }
+        catch (e) {
+            console.error(e);
+            return "Failed to parse Azure Token"
```
Azure AD token

orenbm commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+        }
+        return azureIdentity;
+    }
+</script>
```
missing new line at the end of the file

orenbm reviewed (changes requested):
Thanks @abrahamko !
This tool can really come in handy for SEs and customers!

I left some comments with suggestions to change the language. 

does this project have tests? Do we want to add tests for your tool?

abrahamko commented:
@quincycheng @izgeri  could you please answer @orenbm questions regarding the tests for this project?

orenbm commented:
General note: I noticed that all the logic is implemented in the html file. Is that the best practice? Is it better to have the logic in a js file instead?

izgeri commented:
I know @quincycheng is looking into adding tests - if you have any suggestions for good places to start, or if you have a proposal for what tests maybe could look like here, feel free to share!

orenbm commented:
> I know @quincycheng is looking into adding tests - if you have any suggestions for good places to start, or if you have a proposal for what tests maybe could look like here, feel free to share!

I actually have no experience with JS, maybe @sigalsax can assist?

Tovli commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+            switch (xmsMirid[i]) {
+                case "subscriptions":
+                    azureIdentity.subscriptionId = xmsMirid[i + 1];
+                    i++;
```
why not use continue?

Tovli commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+        var azureIdentity = {};
+
+        var xmsMirid = token.xms_mirid.split('/');
+        for (i = 0; i < xmsMirid.length; i++){
```
I think it is better to mimic the parsing we use in the server. This way you lower the chance to be valid in this tool but fail in the server

abrahamko replied:
As @sigalsax mentioned, we do not depend on exact index in the xms_mirid parsing, so I can't use the [,,,] syntax.
I used `reduce` instead, what do you think of it now?

Tovli reviewed (changes requested):
Great initiative :)
A few engineering work needed
Mainly refactor to be more SOLID and error handling
Good work man 👍

sigalsax commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+
+    function updateCmd() {
+        let output = "";
+        let appId = $("#txtAppId").val();
```
appId -> hostId

sigalsax commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+
+    function getPolicyFromToken(encodedToken, appId) {
+        try {
+            let policy = $("#theTemplate").html();
```
`.html` is a selector where it only takes the first match. So if you have many `theTemplate` elements, it will only take the first HTML element that matches the query. Is that what you want? If not consider using `querySelectorAll`.

abrahamko replied:
Used js template instead.

sigalsax commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+                    i++;
+                    break;
+                case "providers":
+                    azureIdentity.isUserIdentity = xmsMirid[i + 1] === "Microsoft.ManagedIdentity";
```
When developing, we were hesitant to relay on length, indices, etc because if Azure decided to add another property, that would break our design. So I advise you have the same hesitation. 

Take a look how we handled it on the server side. We did something like the following:
``` 
if xms_mirid.providers.include? "Microsoft.ManagedIdentity"
          @token_identity[:user_assigned_identity] = xms_mirid.providers.last
.
.
.
```
So we are checking if `Microsoft.ManagedIdentity` exists anywhere in the providers sub array

sigalsax commented on views/welcome.html:
```diff
@@ -83,8 +84,22 @@ <h2>Integrations</h2>
+          </div>
+        </div>
+      </div>
+
```
extra few lines here

abrahamko replied:
This is how @quincycheng divided between the `div`s in the welcome page.

sigalsax commented:
@abrahamko is your question about which testing framework should be used? or what to test?

The frontend teams in CYBR use Karma and Jasmine
https://codecraft.tv/courses/angular/unit-testing/jasmine-and-karma/ and are pretty satisfied I believe

easiest place to start is testing HTML imo, checking if user gave valid value, etc IMO b/c it is pretty straight forward. what say you @eladkug?

sigalsax commented:
> General note: I noticed that all the logic is implemented in the html file. Is that the best practice? Is it better to have the logic in a js file instead?

good point @orenbm I would actually suggest separating it into it's own JS file to add as a clear separation between view and logic.

eladkug commented:
@sigalsax , i agree with you. Need to check valid/invalid values and add some basic cli test to see its working well.

abrahamko commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,140 @@
+    function parseAzureToken (token) {
+        var azureIdentity = {};
+
+        var xmsMirid = token.xms_mirid.split('/');
```
It will throw, and the error in the catch phrase will be shown.

abrahamko committed:
Fix texts by comments
Move script code to separate file
Wrap js code with anonymous function to prevent littering global scope
Use jwt-decode library for jwt parsing

quincycheng commented:
Thanks for the PR @abrahamko, and great comments & ideas everyone!
It's great to have all of you involved.

I have discussed with @andytinkham earlier about testing.   
Copying @andytinkham , would you mind shedding some light on this?   

I'm not familiar with the process about PR but u've got my green signal.
Please feel free to merge this if everyone agree.

Tovli commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -0,0 +1,71 @@
+        } else {
+            output = getPolicyFromToken(encodedToken, hostId);
+        }
+        $("#output").html(output);
```
do you plan to show here other things besides authn policy? in not then I suggest the id will reflect this

Tovli commented on guides/policy/azure-authenticator/index.html:
```diff
@@ -0,0 +1,45 @@
+        <h3>Policy</h3>
+        <pre class="container border" id="output">
+</pre>
+
```
I would consider adding link to our docs where we explain about azure authn

Tovli commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -0,0 +1,71 @@
+            `;
+        } catch (e) {
+            console.error(e);
+            return "Failed to parse Azure AD Token"
```
Consider showing the error... Think about the flow of a customer that got the "failed to parse" message...
He goes to the repo and open an issue with this text... now what are we need to do to support him?

Tovli commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -0,0 +1,71 @@
+                authn-azure/subscription-id: ${token.subscriptionId}
+                authn-azure/resource-group: ${token.resourceGroup}
+              # The following is optional
+                authn-azure/${token.isUserIdentity ? "user" : "system"}-assigned-identity: ${token.identity}
```
consider building this string before for readability

Tovli commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -0,0 +1,71 @@
+                    azureIdentity.resourceGroup = arr[i + 1];
+                    break;
+                case "providers":
+                    azureIdentity.isUserIdentity = arr.includes("Microsoft.ManagedIdentity", i + 1);
```
you call it here azure identity and afterwards you call it token identity. a bit confusing

Tovli commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    function parseAzureToken(token) {
+        return token.xms_mirid.split('/').reduce((azureIdentity, cur, i, arr) => {
```
From reduce and iteration I actually prefer the iteration as I expect reduce to have more aggregated nature
But I don't feel its critical

Basically I'm not sure you even need to iterate as you look for patterns in a string and can get the next item by using a regex master

Tovli commented on views/welcome.html:
```diff
@@ -83,8 +84,23 @@ <h2>Integrations</h2>
+          <div class="col-6">
+            <Br/>
+            <h2>Azure Authenticator</h2>
+            Generate a policy defining a host that can authenticate with the Azure Authenticator, using an Azure AD token
```
We don't support Azure AD token but Azure managed identity token. It's not the same
You actually have two different Azure AD tokens - B2B token and B2C token... we support neither

Tovli reviewed (changes requested):
Much improvement
Thanks for this

andytinkham commented:
> I have discussed with @andytinkham earlier about testing.
> Copying @andytinkham , would you mind shedding some light on this?

Not exactly sure where you'd like the light, but here's a stab - let me know if you were looking for something else. Tests are one of the things we discussed as being needed to move beyond Community-level certification. In my initial review, I saw 3 big areas to address with tests. First, looking at the generated commands, ideally parsing expected results out of the docs to make sure we stay in sync, though I'm not sure how or if we could actually do that. The most basic case probably runs all the way through executing the commands & verifying the functional components are stood up, but then variants probably just compare the command line generated and the command line expected. 

Area 2 is the UI - restrictions on fields and testing that the combinations of ptions lead to the correct button enablements and/or messages. If this was separated from the view, it would be even better.

Area 3 is addressing error conditions - what happens if fields are left blank, if unsupported values are input, if a container name is specified that matches a container name that already exists, or a password that doesn't meet the complexity requirements. 

In addition to all these, it would be great if we had a way to measure code coverage. I saw @sigalsax mentioned Karma - that definitely has the ability to measure code coverage (I think through plugins), and once we get the Jenkins build running, it would be pretty straightforward to do more with that coverage data. 

Is that what you were looking for, @quincycheng?

abrahamko committed:
Fix comments

Tovli commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -20,52 +20,67 @@
+        return azureIdentity;
+    }
+
+    function findInXmsMirid(name, subscriptionRegex, token) {
```
all "subscriptionRegex" values share the same template... why not give this parameter a better naming? and eliminate the need for "name" parameter?
maybe xmsMiridKey ?
Then you can construct the regex in the function

Tovli reviewed (changes requested):
Minor changes
Great work !!

abrahamko commented on guides/policy/azure-authenticator/parseAzureAuthenticator.js:
```diff
@@ -20,52 +20,67 @@
             output = getPolicyFromToken(encodedToken, hostId);
         }
-        $("#output").html(output);
+        $("#policy").html(output);
```
It was a suggestion from the IDE.
output always has a value other than undefined, anyway.

abrahamko committed:
Fix comments

Tovli reviewed (approved):
LGTM
